lxd/storage/drivers/driver_ceph_volumes.go

```diff
@@ -1558,7 +1607,17 @@ func (d *ceph) MigrateVolume(vol VolumeCopy, conn io.ReadWriteCloser, volSrcArgs
 	}
 
 	if vol.IsVMBlock() {
-		fsVolCopy := NewVolumeCopy(vol.NewVMBlockFilesystemVolume())
+		var fsVolSnapshots []Volume
+		for _, snapshot := range vol.Snapshots {
+			fsVolSnapshots = append(fsVolSnapshots, snapshot.NewVMBlockFilesystemVolume())
+		}
+
+		// Support older versions of LXD that don't yet support optimized volume refresh for Ceph RBD
```

Comment:
I dont follow how older versions are being supported by this - although i am pleased to see it.

Please can you explain how this works in more detail here.

Reply:
Maybe the comment is not at the right place and this leads to confusion.
All I wanted to make clear with this is that the logic for migration slightly differs from refreshing volumes locally.

To support older LXDs, the updated migration procedure has to follow the same steps as the initial migration. Currently the steps are:

Only when refreshing an instance:
1. Send the diff for all VMs filesystem snapshots
2. Send the diff for the VMs filesystem vol

For all other volumes:
1. (3.) Send the diff for all snapshots
2. (4.) Send the diff for the volume

By keeping this order, an older LXD can receive/send the initial migration stream but will fall back to rsync/block transfer for subsequent refreshes. An updated LXD will use the same ordering but will use the RBD migration type also for refreshes.

What this means is that we effectively change the logic how we identify what has to be sent/received but send/receive the stream in the right order so that an older LXD can still recognize it.

Reply:
Thanks, this is what the comment would benefit from:

> What this means is that we effectively change the logic how we identify what has to be sent/received but send/receive the stream in the right order so that an older LXD can still recognize it.

Reply:
Fixed.